<a href="https://colab.research.google.com/github/Gabriel-data/Colab/blob/main/3_PREPARA_DADOS_curso_Est_Dados_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Preparação, organização e estruturação dos dados.

## Conhecendo o dataset (A17)

Etapa do conteúdo e ações de ajustes de estrutura do dataset.
Anexo em: https://drive.google.com/file/d/1k40uHWBuhdZlGVz1yy_X2x7TCrzg4_rh/view?usp=sharing


PROJETO ANÁLISE ESTATÍSTICA DO ENEM 2019

Este projeto analisa os resultados do ENEM no estado de São Paulo do ano de 2019.

Os dados foram extraídos do site do INEP:

http://inep.gov.br/microdados

## Preparação, organização e estruturação dos dados: Parte 1 (A18)

In [1]:
### Montar o Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Importar as bibliotecas
import numpy as np
import pandas as pd

In [3]:
# CARREGAR O DATASET (1-2)

dados = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Arquivos/curso_Est_Dados_Python_3-2024/microdados_enem_2019_sp.csv',sep=';', encoding='iso-8859-1')
# encoding: codificação de caracteres, normalmente utiliza-se o iso-8859-1, utf-8, latin-1)

In [ ]:
# # CARREGAR O DATASET (2-2) OBSERVAÇÕES:

# Revisar a cód. de caracteres da tabela de dados:
# sep=';' or ','
# encoding='iso-8859-1', utf-8, latin-1

### Extração e Limpeza dos dados

In [4]:
pd.options.display.max_columns = 44

In [5]:
# Apresenta o dataset
dados.head()

NU_INSCRICAO  CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  \
0     1,90E+11                  3546801            Santa Isabel   
1     1,90E+11                  3538204             Pinhalzinho   
2     1,90E+11                  3550308               São Paulo   
3     1,90E+11                  3548708   São Bernardo do Campo   
4     1,90E+11                  3549409    São Joaquim da Barra   

   CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0                35               SP        22       M                1   
1                35               SP        19       M                1   
2                35               SP        17       F                1   
3                35               SP        19       M                1   
4                35               SP        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  CO_MUNICIPIO_NASCIMENTO  \
0            3                 1                1506500.0   
1            3                 1                1503804.0   
2            3                 1                1501402.0   
3            3                 1                3548708.0   
4            1                 1                3549409.0   

  NO_MUNICIPIO_NASCIMENTO  CO_UF_NASCIMENTO SG_UF_NASCIMENTO  TP_ST_CONCLUSAO  \
0    Santa Izabel do Pará              15.0               PA                1   
1                 Jacundá              15.0               PA                1   
2                   Belém              15.0               PA                2   
3   São Bernardo do Campo              35.0               SP                2   
4    São Joaquim da Barra              35.0               SP                1   

   TP_ANO_CONCLUIU  TP_ESCOLA  TP_ENSINO  IN_TREINEIRO   CO_ESCOLA  \
0                4          1        1.0             0         NaN   
1                2          1        1.0             0         NaN   
2                0          3        1.0             0  15567761.0   
3                0          2        1.0             0  35904958.0   
4                3          1        1.0             0         NaN   

   CO_MUNICIPIO_ESC       NO_MUNICIPIO_ESC  CO_UF_ESC SG_UF_ESC  \
0               NaN                    NaN        NaN       NaN   
1               NaN                    NaN        NaN       NaN   
2         1501402.0                  Belém       15.0        PA   
3         3548708.0  São Bernardo do Campo       35.0        SP   
4               NaN                    NaN        NaN       NaN   

   TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  TP_SIT_FUNC_ESC  \
0                     NaN                 NaN              NaN   
1                     NaN                 NaN              NaN   
2                     4.0                 1.0              1.0   
3                     2.0                 1.0              1.0   
4                     NaN                 NaN              NaN   

   TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  TP_PRESENCA_MT  NU_NOTA_CN  \
0               1               1               1               1      5646.0   
1               0               0               0               0         NaN   
2               1               1               1               1      5762.0   
3               1               1               1               1      4022.0   
4               0               0               0               0         NaN   

   NU_NOTA_CH  NU_NOTA_LC  NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  \
0      5858.0      5929.0      6002.0          0                1.0   
1         NaN         NaN         NaN          1                NaN   
2      6411.0      6343.0      7318.0          0                1.0   
3      4248.0      5096.0      3932.0          1                1.0   
4         NaN         NaN         NaN          1                NaN   

   NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0          200.0          180.0          200.0          200.0   

In [6]:
#O número total de inscritos em São Paulo foi de 813772, sendo de 15,97% de todo o Brasil que foi de 5095308 inscritos.
dados.shape

(813772, 44)

In [7]:
# Análise dos tipos de atributos.
# object: strings
# int64: inteiros
# float64: reais
# complex: complexos
dados.dtypes

NU_INSCRICAO                object
CO_MUNICIPIO_RESIDENCIA      int64
NO_MUNICIPIO_RESIDENCIA     object
CO_UF_RESIDENCIA             int64
SG_UF_RESIDENCIA            object
NU_IDADE                     int64
TP_SEXO                     object
TP_ESTADO_CIVIL              int64
TP_COR_RACA                  int64
TP_NACIONALIDADE             int64
CO_MUNICIPIO_NASCIMENTO    float64
NO_MUNICIPIO_NASCIMENTO     object
CO_UF_NASCIMENTO           float64
SG_UF_NASCIMENTO            object
TP_ST_CONCLUSAO              int64
TP_ANO_CONCLUIU              int64
TP_ESCOLA                    int64
TP_ENSINO                  float64
IN_TREINEIRO                 int64
CO_ESCOLA                  float64
CO_MUNICIPIO_ESC           float64
NO_MUNICIPIO_ESC            object
CO_UF_ESC                  float64
SG_UF_ESC                   object
TP_DEPENDENCIA_ADM_ESC     float64
TP_LOCALIZACAO_ESC         float64
TP_SIT_FUNC_ESC            float64
TP_PRESENCA_CN               int64
TP_PRESENCA_CH      

In [10]:
# Refazer um novo dataset e retirar colinas específicas

dados1 = dados.drop(columns=['CO_MUNICIPIO_RESIDENCIA'])

In [11]:
dados1 = dados1.drop(columns=['CO_UF_RESIDENCIA','SG_UF_RESIDENCIA','CO_MUNICIPIO_NASCIMENTO'], )

In [13]:
dados1 = dados1.drop(columns=['NO_MUNICIPIO_NASCIMENTO','CO_UF_NASCIMENTO','SG_UF_NASCIMENTO',
                              'TP_ANO_CONCLUIU','TP_ENSINO','CO_MUNICIPIO_ESC','CO_UF_ESC','SG_UF_ESC'],)

In [15]:
dados1 = dados1.drop(columns=['TP_DEPENDENCIA_ADM_ESC','TP_LOCALIZACAO_ESC','TP_SIT_FUNC_ESC'],)

In [16]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0     1,90E+11            Santa Isabel        22       M                1   
1     1,90E+11             Pinhalzinho        19       M                1   
2     1,90E+11               São Paulo        17       F                1   
3     1,90E+11   São Bernardo do Campo        19       M                1   
4     1,90E+11    São Joaquim da Barra        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ESCOLA  IN_TREINEIRO  \
0            3                 1                1          1             0   
1            3                 1                1          1             0   
2            3                 1                2          3             0   
3            3                 1                2          2             0   
4            1                 1                1          1             0   

    CO_ESCOLA       NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  \
0         NaN                    NaN               1               1   
1         NaN                    NaN               0               0   
2  15567761.0                  Belém               1               1   
3  35904958.0  São Bernardo do Campo               1               1   
4         NaN                    NaN               0               0   

   TP_PRESENCA_LC  TP_PRESENCA_MT  NU_NOTA_CN  NU_NOTA_CH  NU_NOTA_LC  \
0               1               1      5646.0      5858.0      5929.0   
1               0               0         NaN         NaN         NaN   
2               1               1      5762.0      6411.0      6343.0   
3               1               1      4022.0      4248.0      5096.0   
4               0               0         NaN         NaN         NaN   

   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  NU_NOTA_COMP1  NU_NOTA_COMP2  \
0      6002.0          0                1.0          160.0          200.0   
1         NaN          1                NaN            NaN            NaN   
2      7318.0          0                1.0          160.0          120.0   
3      3932.0          1                1.0           80.0          160.0   
4         NaN          1                NaN            NaN            NaN   

   NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO Q025  
0          180.0          200.0          200.0            940.0    B  
1            NaN            NaN            NaN              NaN    A  
2          200.0          200.0          200.0            880.0    A  
3          120.0          100.0          100.0            560.0    B  
4            NaN            NaN            NaN              NaN    B

In [17]:
dados1.shape

(813772, 29)

## Preparação, organização e estruturação dos dados: Parte 2 (A19)

In [18]:
# Identificando os erros atráves do conhecimento sobre dados e leitura de informações
dados.head()

NU_INSCRICAO  CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  \
0     1,90E+11                  3546801            Santa Isabel   
1     1,90E+11                  3538204             Pinhalzinho   
2     1,90E+11                  3550308               São Paulo   
3     1,90E+11                  3548708   São Bernardo do Campo   
4     1,90E+11                  3549409    São Joaquim da Barra   

   CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0                35               SP        22       M                1   
1                35               SP        19       M                1   
2                35               SP        17       F                1   
3                35               SP        19       M                1   
4                35               SP        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  CO_MUNICIPIO_NASCIMENTO  \
0            3                 1                1506500.0   
1            3                 1                1503804.0   
2            3                 1                1501402.0   
3            3                 1                3548708.0   
4            1                 1                3549409.0   

  NO_MUNICIPIO_NASCIMENTO  CO_UF_NASCIMENTO SG_UF_NASCIMENTO  TP_ST_CONCLUSAO  \
0    Santa Izabel do Pará              15.0               PA                1   
1                 Jacundá              15.0               PA                1   
2                   Belém              15.0               PA                2   
3   São Bernardo do Campo              35.0               SP                2   
4    São Joaquim da Barra              35.0               SP                1   

   TP_ANO_CONCLUIU  TP_ESCOLA  TP_ENSINO  IN_TREINEIRO   CO_ESCOLA  \
0                4          1        1.0             0         NaN   
1                2          1        1.0             0         NaN   
2                0          3        1.0             0  15567761.0   
3                0          2        1.0             0  35904958.0   
4                3          1        1.0             0         NaN   

   CO_MUNICIPIO_ESC       NO_MUNICIPIO_ESC  CO_UF_ESC SG_UF_ESC  \
0               NaN                    NaN        NaN       NaN   
1               NaN                    NaN        NaN       NaN   
2         1501402.0                  Belém       15.0        PA   
3         3548708.0  São Bernardo do Campo       35.0        SP   
4               NaN                    NaN        NaN       NaN   

   TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  TP_SIT_FUNC_ESC  \
0                     NaN                 NaN              NaN   
1                     NaN                 NaN              NaN   
2                     4.0                 1.0              1.0   
3                     2.0                 1.0              1.0   
4                     NaN                 NaN              NaN   

   TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  TP_PRESENCA_MT  NU_NOTA_CN  \
0               1               1               1               1      5646.0   
1               0               0               0               0         NaN   
2               1               1               1               1      5762.0   
3               1               1               1               1      4022.0   
4               0               0               0               0         NaN   

   NU_NOTA_CH  NU_NOTA_LC  NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  \
0      5858.0      5929.0      6002.0          0                1.0   
1         NaN         NaN         NaN          1                NaN   
2      6411.0      6343.0      7318.0          0                1.0   
3      4248.0      5096.0      3932.0          1                1.0   
4         NaN         NaN         NaN          1                NaN   

   NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0          200.0          180.0          200.0          200.0   

Corrigindo erros nas notas

In [19]:
# localiza em todas as linhas os valores e realiza a divisão /10
dados1.loc[:,'NU_NOTA_CN'] /= 10

In [21]:
dados1.head(5)

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0     1,90E+11            Santa Isabel        22       M                1   
1     1,90E+11             Pinhalzinho        19       M                1   
2     1,90E+11               São Paulo        17       F                1   
3     1,90E+11   São Bernardo do Campo        19       M                1   
4     1,90E+11    São Joaquim da Barra        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ESCOLA  IN_TREINEIRO  \
0            3                 1                1          1             0   
1            3                 1                1          1             0   
2            3                 1                2          3             0   
3            3                 1                2          2             0   
4            1                 1                1          1             0   

    CO_ESCOLA       NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  \
0         NaN                    NaN               1               1   
1         NaN                    NaN               0               0   
2  15567761.0                  Belém               1               1   
3  35904958.0  São Bernardo do Campo               1               1   
4         NaN                    NaN               0               0   

   TP_PRESENCA_LC  TP_PRESENCA_MT  NU_NOTA_CN  NU_NOTA_CH  NU_NOTA_LC  \
0               1               1       564.6      5858.0      5929.0   
1               0               0         NaN         NaN         NaN   
2               1               1       576.2      6411.0      6343.0   
3               1               1       402.2      4248.0      5096.0   
4               0               0         NaN         NaN         NaN   

   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  NU_NOTA_COMP1  NU_NOTA_COMP2  \
0      6002.0          0                1.0          160.0          200.0   
1         NaN          1                NaN            NaN            NaN   
2      7318.0          0                1.0          160.0          120.0   
3      3932.0          1                1.0           80.0          160.0   
4         NaN          1                NaN            NaN            NaN   

   NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO Q025  
0          180.0          200.0          200.0            940.0    B  
1            NaN            NaN            NaN              NaN    A  
2          200.0          200.0          200.0            880.0    A  
3          120.0          100.0          100.0            560.0    B  
4            NaN            NaN            NaN              NaN    B

In [22]:
dados1.loc[:,'NU_NOTA_CH'] /= 10
dados1.loc[:,'NU_NOTA_LC'] /= 10
dados1.loc[:,'NU_NOTA_MT'] /= 10

In [23]:
dados1.head(5)

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0     1,90E+11            Santa Isabel        22       M                1   
1     1,90E+11             Pinhalzinho        19       M                1   
2     1,90E+11               São Paulo        17       F                1   
3     1,90E+11   São Bernardo do Campo        19       M                1   
4     1,90E+11    São Joaquim da Barra        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ESCOLA  IN_TREINEIRO  \
0            3                 1                1          1             0   
1            3                 1                1          1             0   
2            3                 1                2          3             0   
3            3                 1                2          2             0   
4            1                 1                1          1             0   

    CO_ESCOLA       NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  \
0         NaN                    NaN               1               1   
1         NaN                    NaN               0               0   
2  15567761.0                  Belém               1               1   
3  35904958.0  São Bernardo do Campo               1               1   
4         NaN                    NaN               0               0   

   TP_PRESENCA_LC  TP_PRESENCA_MT  NU_NOTA_CN  NU_NOTA_CH  NU_NOTA_LC  \
0               1               1       564.6       585.8       592.9   
1               0               0         NaN         NaN         NaN   
2               1               1       576.2       641.1       634.3   
3               1               1       402.2       424.8       509.6   
4               0               0         NaN         NaN         NaN   

   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  NU_NOTA_COMP1  NU_NOTA_COMP2  \
0       600.2          0                1.0          160.0          200.0   
1         NaN          1                NaN            NaN            NaN   
2       731.8          0                1.0          160.0          120.0   
3       393.2          1                1.0           80.0          160.0   
4         NaN          1                NaN            NaN            NaN   

   NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO Q025  
0          180.0          200.0          200.0            940.0    B  
1            NaN            NaN            NaN              NaN    A  
2          200.0          200.0          200.0            880.0    A  
3          120.0          100.0          100.0            560.0    B  
4            NaN            NaN            NaN              NaN    B

Renomeando Variáveis (colunas)

In [24]:
dados1 = dados1.rename(columns={'NU_NOTA_REDACAO': 'NOTA_REDACAO'})

In [25]:
dados1 = dados1.rename(columns={'NU_NOTA_CN': 'NOTA_CN','NU_NOTA_CH': 'NOTA_CH',
                                'NU_NOTA_LC': 'NOTA_LC','NU_NOTA_MT': 'NOTA_MT'})

In [26]:
dados1 = dados1.rename(columns={'NU_NOTA_COMP1': 'COMP1','NU_NOTA_COMP2': 'COMP2',
                                'NU_NOTA_COMP3': 'COMP3','NU_NOTA_COMP4': 'COMP4',
                                'NU_NOTA_COMP5':'COMP5'})

In [27]:
dados1 = dados1.rename(columns={'NU_IDADE': 'IDADE','TP_SEXO': 'SEXO',
                                'TP_COR_RACA': 'RACA','Q025': 'INTERNET',
                                'TP_ESCOLA':'ESCOLA'})

In [30]:
dados1 = dados1.rename(columns={'TP_NACIONALIDADE': 'NACIONALIDADE'})

In [31]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL  RACA  \
0     1,90E+11            Santa Isabel     22    M                1     3   
1     1,90E+11             Pinhalzinho     19    M                1     3   
2     1,90E+11               São Paulo     17    F                1     3   
3     1,90E+11   São Bernardo do Campo     19    M                1     3   
4     1,90E+11    São Joaquim da Barra     37    M                2     1   

   NACIONALIDADE  TP_ST_CONCLUSAO  ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1       1             0         NaN   
1              1                1       1             0         NaN   
2              1                2       3             0  15567761.0   
3              1                2       2             0  35904958.0   
4              1                1       1             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
1                    NaN               0               0               0   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
4                    NaN               0               0               0   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT  TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2          0   
1               0      NaN      NaN      NaN      NaN          1   
2               1    576.2    641.1    634.3    731.8          0   
3               1    402.2    424.8    509.6    393.2          1   
4               0      NaN      NaN      NaN      NaN          1   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0        B  
1                NaN    NaN    NaN    NaN    NaN    NaN           NaN        A  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0        A  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0        B  
4                NaN    NaN    NaN    NaN    NaN    NaN           NaN        B

Renomeando registros

In [32]:
dados1["RACA"] = dados1["RACA"].replace({0:"nao_declarado", 1:"branca", 2:"preta", 3:"parda", 4:"amarela", 5:"indigena"})

# onde: ["RACA"] = coluna
# onde: {0:, 1:, 2:...} = registros que anteriormente estavam indicados apenas por números.

In [33]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   NACIONALIDADE  TP_ST_CONCLUSAO  ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1       1             0         NaN   
1              1                1       1             0         NaN   
2              1                2       3             0  15567761.0   
3              1                2       2             0  35904958.0   
4              1                1       1             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
1                    NaN               0               0               0   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
4                    NaN               0               0               0   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT  TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2          0   
1               0      NaN      NaN      NaN      NaN          1   
2               1    576.2    641.1    634.3    731.8          0   
3               1    402.2    424.8    509.6    393.2          1   
4               0      NaN      NaN      NaN      NaN          1   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0        B  
1                NaN    NaN    NaN    NaN    NaN    NaN           NaN        A  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0        A  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0        B  
4                NaN    NaN    NaN    NaN    NaN    NaN           NaN        B

In [ ]:
# OUTRA MANEIRA
# dados1['RACA'] = dados1['RACA'].replace([0,1,2,3,4,5], ['Não Declarado','Branca','Preta','Parda','Amarela','Indígena'])

In [34]:
dados1["TP_LINGUA"] = dados1["TP_LINGUA"].replace({0:"Inglês", 1:"Espanhol"})

In [35]:
dados1["ESCOLA"] = dados1["ESCOLA"].replace({1:"não_respondeu", 2:"pública", 3:"privada", 4:"exterior"})

In [36]:
dados1["INTERNET"] = dados1["INTERNET"].replace({"A":"Não", "B":"Sim"})

In [37]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1  não_respondeu             0         NaN   
1              1                1  não_respondeu             0         NaN   
2              1                2        privada             0  15567761.0   
3              1                2        pública             0  35904958.0   
4              1                1  não_respondeu             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
1                    NaN               0               0               0   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
4                    NaN               0               0               0   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2    Inglês   
1               0      NaN      NaN      NaN      NaN  Espanhol   
2               1    576.2    641.1    634.3    731.8    Inglês   
3               1    402.2    424.8    509.6    393.2  Espanhol   
4               0      NaN      NaN      NaN      NaN  Espanhol   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0      Sim  
1                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Não  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0      Não  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0      Sim  
4                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Sim

## Análise Exploratória e Transformação dos dados

**Análise das idades**

In [39]:
# contagem dos valores da coluna IDADE
dados1['IDADE'].value_counts()

17    219313
18    147548
19     81236
16     59316
20     53009
       ...  
86         1
4          1
3          1
91         1
2          1
Name: IDADE, Length: 80, dtype: int64

In [40]:
# contagem por Ordem (.sort_index())

dados1['IDADE'].value_counts().sort_index()

2     1
3     1
4     1
5     2
6     2
     ..
80    2
81    3
82    2
86    1
91    1
Name: IDADE, Length: 80, dtype: int64

Alguns valores estranhos e curiosos: idades menores que 12 anos e maiores de 80 anos.

In [41]:
# Query (1-2) para separar e entender valores suspeitos / valores < 12

menores_12 = dados1.query('IDADE < 12')['NO_MUNICIPIO_RESIDENCIA'].value_counts()

In [43]:
menores_12

São Paulo                 5
Tambaú                    1
Ribeirão Preto            1
Bálsamo                   1
São Bernardo do Campo     1
Embu das Artes            1
Queiroz                   1
Santana de Parnaíba       1
Vargem Grande Paulista    1
Auriflama                 1
Araçatuba                 1
Campinas                  1
Name: NO_MUNICIPIO_RESIDENCIA, dtype: int64

In [44]:
# # Filtrando parâmetros de dadtaset para outro dataset

dados_maiores_11 = dados1.loc[dados1.IDADE > 11]

In [45]:
dados_maiores_11.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1  não_respondeu             0         NaN   
1              1                1  não_respondeu             0         NaN   
2              1                2        privada             0  15567761.0   
3              1                2        pública             0  35904958.0   
4              1                1  não_respondeu             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
1                    NaN               0               0               0   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
4                    NaN               0               0               0   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2    Inglês   
1               0      NaN      NaN      NaN      NaN  Espanhol   
2               1    576.2    641.1    634.3    731.8    Inglês   
3               1    402.2    424.8    509.6    393.2  Espanhol   
4               0      NaN      NaN      NaN      NaN  Espanhol   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0      Sim  
1                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Não  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0      Não  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0      Sim  
4                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Sim

In [46]:
dados_maiores_11['IDADE'].value_counts().sort_index()

12        6
13       29
14      355
15    11583
16    59316
      ...  
80        2
81        3
82        2
86        1
91        1
Name: IDADE, Length: 73, dtype: int64

In [48]:
# Configura o Pandas para observar 80 linhas
pd.set_option('display.max_rows', 80)

**Treineiros**

In [49]:
# Contar quantos Treineiros fizeram a prova do Enen

dados_maiores_11.query('IN_TREINEIRO == 1')['IN_TREINEIRO'].value_counts()

1    93988
Name: IN_TREINEIRO, dtype: int64

In [52]:
treineiros = dados_maiores_11.loc[dados_maiores_11.IN_TREINEIRO == 1]

In [53]:
treineiros.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
20      1,90E+11                  Santos     15    F                1   parda   
21      1,90E+11                Panorama     16    F                1   parda   
106     1,90E+11               São Paulo     16    F                1  branca   
109     1,90E+11        Águas de Lindóia     16    M                1  branca   
159     1,90E+11        Águas de Lindóia     15    M                1  branca   

     NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO  CO_ESCOLA  \
20               1                3  não_respondeu             1        NaN   
21               1                3  não_respondeu             1        NaN   
106              1                3  não_respondeu             1        NaN   
109              1                3  não_respondeu             1        NaN   
159              1                3  não_respondeu             1        NaN   

    NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
20               NaN               1               1               1   
21               NaN               0               0               0   
106              NaN               1               1               1   
109              NaN               0               0               0   
159              NaN               0               0               0   

     TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT TP_LINGUA  \
20                1    455.4    506.7    480.4    573.2    Inglês   
21                0      NaN      NaN      NaN      NaN  Espanhol   
106               1    424.5    517.7    489.6    532.6    Inglês   
109               0      NaN      NaN      NaN      NaN  Espanhol   
159               0      NaN      NaN      NaN      NaN    Inglês   

     TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO  \
20                 1.0  140.0  120.0  120.0  100.0   80.0         560.0   
21                 NaN    NaN    NaN    NaN    NaN    NaN           NaN   
106                1.0  120.0  120.0  120.0  100.0   80.0         540.0   
109                NaN    NaN    NaN    NaN    NaN    NaN           NaN   
159                NaN    NaN    NaN    NaN    NaN    NaN           NaN   

    INTERNET  
20       Sim  
21       Não  
106      Sim  
109      Sim  
159      Sim

In [54]:
treineiros.to_csv('treineiros_enem_2019_sp.csv', encoding = 'iso-8859-1', index = False)

In [55]:
# Separação entre os treineiros e os vestibulandos
vestibulandos = dados_maiores_11.loc[dados_maiores_11.IN_TREINEIRO == 0]

In [61]:
vestibulandos.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
1     1,90E+11             Pinhalzinho     19    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
4     1,90E+11    São Joaquim da Barra     37    M                2  branca   

   NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1  não_respondeu             0         NaN   
1              1                1  não_respondeu             0         NaN   
2              1                2        privada             0  15567761.0   
3              1                2        pública             0  35904958.0   
4              1                1  não_respondeu             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
1                    NaN               0               0               0   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
4                    NaN               0               0               0   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2    Inglês   
1               0      NaN      NaN      NaN      NaN  Espanhol   
2               1    576.2    641.1    634.3    731.8    Inglês   
3               1    402.2    424.8    509.6    393.2  Espanhol   
4               0      NaN      NaN      NaN      NaN  Espanhol   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0      Sim  
1                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Não  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0      Não  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0      Sim  
4                NaN    NaN    NaN    NaN    NaN    NaN           NaN      Sim

**Análise das presenças**

In [57]:
vestibulandos['TP_PRESENCA_CN'].value_counts()

1    500805
0    218693
2       270
Name: TP_PRESENCA_CN, dtype: int64

In [58]:
vestibulandos['TP_PRESENCA_CH'].value_counts()

1    533142
0    185969
2       657
Name: TP_PRESENCA_CH, dtype: int64

In [59]:
vestibulandos['TP_PRESENCA_LC'].value_counts()

1    533142
0    185969
2       657
Name: TP_PRESENCA_LC, dtype: int64

In [64]:
vestibulandos['TP_PRESENCA_MT'].value_counts()

1    500805
0    218693
2       270
Name: TP_PRESENCA_MT, dtype: int64

RESUMO DA PRESENÇA NAS PROVAS:

Ciências da Natureza: 500805 presentes, 218693 faltaram e 270 eliminados.

Ciências Humanas: 533142 presentes, 185969 faltaram e 657 eliminados.

Linguagens e Códigos: 533142 presentes, 185969 faltaram e 657 eliminados.

Matemática: 500805 presentes, 218693 faltaram e 270 eliminados.

In [65]:
vestibulandos['TP_STATUS_REDACAO'].value_counts().sort_index()

1.0    519763
2.0       913
3.0      1417
4.0      5513
6.0      3626
7.0       543
8.0       828
9.0       539
Name: TP_STATUS_REDACAO, dtype: int64

**Valores Ausentes**

In [66]:
# Identificar valores ausentes

vestibulandos['IDADE'].isnull().sum()

0

In [67]:
# Criar uma variável

provas = ['NOTA_REDACAO','NOTA_CN','NOTA_MT','NOTA_LC','NOTA_CH']

In [68]:
# Registros NAN nas notas
vestibulandos[provas].isnull().sum()

NOTA_REDACAO    186626
NOTA_CN         218963
NOTA_MT         218963
NOTA_LC         186626
NOTA_CH         186626
dtype: int64

In [70]:
# Criar um dataset apenas com os vestibulandos classificados

vestibulandos_classificados = vestibulandos.loc[vestibulandos.TP_PRESENCA_CH == 1]
vestibulandos_classificados = vestibulandos_classificados.loc[vestibulandos_classificados.TP_PRESENCA_MT == 1]

In [71]:
vestibulandos_classificados['TP_PRESENCA_MT'].value_counts()

1    499940
Name: TP_PRESENCA_MT, dtype: int64

In [72]:
vestibulandos_classificados['TP_PRESENCA_CH'].value_counts()

1    499940
Name: TP_PRESENCA_CH, dtype: int64

In [73]:
vestibulandos_classificados['TP_PRESENCA_CN'].value_counts()

1    499940
Name: TP_PRESENCA_CN, dtype: int64

In [74]:
vestibulandos_classificados['TP_PRESENCA_LC'].value_counts()

1    499940
Name: TP_PRESENCA_LC, dtype: int64

In [75]:
vestibulandos_classificados[provas].isnull().sum()

NOTA_REDACAO    0
NOTA_CN         0
NOTA_MT         0
NOTA_LC         0
NOTA_CH         0
dtype: int64

** Notas zeros na Redação**

In [77]:
from collections import Counter
dict(Counter(vestibulandos_classificados['NOTA_REDACAO']))

{940.0: 4688,
 880.0: 7676,
 560.0: 35741,
 580.0: 33062,
 540.0: 25538,
 900.0: 5921,
 520.0: 24551,
 500.0: 17102,
 680.0: 19520,
 660.0: 18451,
 760.0: 12859,
 860.0: 6502,
 480.0: 19617,
 640.0: 27623,
 720.0: 15214,
 600.0: 37197,
 820.0: 7810,
 340.0: 5464,
 0.0: 9181,
 920.0: 6540,
 780.0: 9236,
 400.0: 7232,
 740.0: 10931,
 320.0: 8113,
 840.0: 9047,
 800.0: 10677,
 700.0: 13502,
 460.0: 13045,
 360.0: 7023,
 620.0: 28059,
 980.0: 978,
 440.0: 11950,
 420.0: 6555,
 380.0: 4453,
 280.0: 4734,
 300.0: 5418,
 960.0: 3057,
 200.0: 312,
 260.0: 2634,
 180.0: 105,
 240.0: 1841,
 220.0: 664,
 40.0: 1,
 140.0: 20,
 160.0: 65,
 120.0: 16,
 100.0: 5,
 60.0: 4,
 1000.0: 5,
 80.0: 1}

In [78]:
vestibulandos_classificados.query('NOTA_REDACAO == 0')['NOTA_REDACAO'].value_counts()

0.0    9181
Name: NOTA_REDACAO, dtype: int64

In [79]:
vestibulandos_classificados.query('COMP1 == 0')['COMP1'].value_counts()

0.0    9188
Name: COMP1, dtype: int64

In [81]:
vestibulandos_classificados.query('COMP2 == 0')['COMP2'].value_counts()

0.0    9181
Name: COMP2, dtype: int64

In [82]:
vestibulandos_classificados.query('COMP3 == 0')['COMP3'].value_counts()

0.0    9197
Name: COMP3, dtype: int64

In [83]:
vestibulandos_classificados.query('COMP4 == 0')['COMP4'].value_counts()

0.0    9194
Name: COMP4, dtype: int64

In [84]:
vestibulandos_classificados.query('COMP5 == 0')['COMP5'].value_counts()

0.0    73330
Name: COMP5, dtype: int64

NOTAS ZEROS EM REDAÇÃO:

REDAÇÃO (final): 9181 notas zeros devido a não cumprimento das regras ou tirou zero em todos os cinco componentes.

COMPONENTE 1(Demonstrar domínio da modalidade escrita formal da Língua Portuguesa.): 9188 notas zeros somente neste item.

COMPONENTE 2 (Compreender a proposta de redação e aplicar conceitos das várias áreas de conhecimento para desenvolver o tema, dentro dos limites estruturais do texto dissertativo-argumentativo em prosa): 9181.

COMPONENTE 3 (Selecionar, relacionar, organizar e interpretar informações, fatos, opiniões e argumentos em defesa de um ponto de vista): 9197 notas zeros somente neste item.

COMPONENTE 4 (Demonstrar conhecimento dos mecanismos linguísticos necessários para a construção da argumentação.): 9194 notas zeros somente neste item.

COMPONENTE 5 (Elaborar proposta de intervenção para o problema abordado, respeitando os direitos humanos): 73330 notas zeros somente neste item.


In [85]:
vestibulandos_classificados.query('NOTA_CH == 0')['NOTA_CH'].value_counts()

0.0    343
Name: NOTA_CH, dtype: int64

In [86]:
vestibulandos_classificados.query('NOTA_CN == 0')['NOTA_CN'].value_counts()

0.0    36
Name: NOTA_CN, dtype: int64

In [87]:
vestibulandos_classificados.query('NOTA_LC == 0')['NOTA_LC'].value_counts()

0.0    135
Name: NOTA_LC, dtype: int64

In [88]:
vestibulandos_classificados.query('NOTA_MT == 0')['NOTA_MT'].value_counts()

0.0    56
Name: NOTA_MT, dtype: int64

In [89]:
vestibulandos_classificados.query('NOTA_REDACAO == 0')['NOTA_REDACAO'].value_counts()

0.0    9181
Name: NOTA_REDACAO, dtype: int64

NOTAS ZEROS DOS INSCRITOS CONCORRENTES AO VESTIBULAR:

Matemática = 56

Ciências da Natureza = 36

Linguagens e Códigos = 135

Ciências Humanas = 343

Redação = 9181


In [90]:
vestibulandos_classificados.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  IDADE SEXO  TP_ESTADO_CIVIL    RACA  \
0     1,90E+11            Santa Isabel     22    M                1   parda   
2     1,90E+11               São Paulo     17    F                1   parda   
3     1,90E+11   São Bernardo do Campo     19    M                1   parda   
5     1,90E+11              Embu-Guaçu     17    F                1   parda   
8     1,90E+11               Guarulhos     23    F                1  branca   

   NACIONALIDADE  TP_ST_CONCLUSAO         ESCOLA  IN_TREINEIRO   CO_ESCOLA  \
0              1                1  não_respondeu             0         NaN   
2              1                2        privada             0  15567761.0   
3              1                2        pública             0  35904958.0   
5              1                2        pública             0         NaN   
8              1                1  não_respondeu             0         NaN   

        NO_MUNICIPIO_ESC  TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  \
0                    NaN               1               1               1   
2                  Belém               1               1               1   
3  São Bernardo do Campo               1               1               1   
5                    NaN               1               1               1   
8                    NaN               1               1               1   

   TP_PRESENCA_MT  NOTA_CN  NOTA_CH  NOTA_LC  NOTA_MT TP_LINGUA  \
0               1    564.6    585.8    592.9    600.2    Inglês   
2               1    576.2    641.1    634.3    731.8    Inglês   
3               1    402.2    424.8    509.6    393.2  Espanhol   
5               1    399.6    426.8    487.2    419.2    Inglês   
8               1    625.7    646.5    629.3    796.1    Inglês   

   TP_STATUS_REDACAO  COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO INTERNET  
0                1.0  160.0  200.0  180.0  200.0  200.0         940.0      Sim  
2                1.0  160.0  120.0  200.0  200.0  200.0         880.0      Não  
3                1.0   80.0  160.0  120.0  100.0  100.0         560.0      Sim  
5                1.0  120.0  120.0  120.0  120.0   80.0         560.0      Sim  
8                1.0  160.0  200.0  180.0  180.0  160.0         880.0      Sim

In [91]:
vestibulandos_classificados = vestibulandos_classificados.drop(columns=['IN_TREINEIRO'], )

In [92]:
vestibulandos_classificados.shape

(499940, 28)

## **Salvando (Exportando) o Dataframe Tratado**

In [93]:
vestibulandos_classificados.to_csv('enem_2019_tratado.csv', encoding = 'iso-8859-1', index = False)

### **REFERÊNCIAS**

PROJETO ANÁLISE ESTATÍSTICA DO ENEM 2019

Base geral:
https://colab.research.google.com/drive/1ncjWndPVkD-HVSqN1xdyPFFRacTIvF0D#scrollTo=FOFTR-sMvNsK&line=2&uniqifier=1

Dicionário Microdados Enen 2019 (XLS):

https://colab.research.google.com/drive/1ncjWndPVkD-HVSqN1xdyPFFRacTIvF0D#scrollTo=oeSMcWvdwHsA&line=3&uniqifier=1

Microdados Enen 2019 (CSV):

https://colab.research.google.com/drive/1ncjWndPVkD-HVSqN1xdyPFFRacTIvF0D#scrollTo=jjeY2AhNwATU&line=3&uniqifier=1